libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+        {
+            It tmp = rhs;
+            difference_type res = 0;
+            while (tmp++ != *this)
+            {
+                ++res;
+            }
+            return res;
+        }
+
+        reference operator*() const
+        {
+            return *m_iter;
+        }
+
+        pointer operator->() const
+        {
+            return m_iter.operator->();
+        }
+
+        bool equal(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter == rhs.m_iter;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, bool>
+        less_than(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter < rhs.m_iter;
+        }
+
+    private:
+
+        void advance(difference_type n)
+        {
+            while (m_iter != m_end && n != 0)
+            {
+                ++(*this), --n;
+            }
+        }
+
+        void step_back(difference_type n)
+        {
+            while (m_iter != m_begin_limit && n != 0)
+            {
+                --(*this), --n;
+            }
+        }
+
+        void satisfy_predicate()
+        {
+            while (m_iter != m_end && !m_pred(*m_iter))
+            {
+                ++m_iter;
+            }
+        }
+
+        Predicate m_pred;
+        Iterator m_iter;
+        Iterator m_begin_limit;
+        Iterator m_end;
+    };
+
+    template <class Predicate, class Iterator>
+    bool operator==(const filter_iterator<Predicate, Iterator>& lhs,
+                    const filter_iterator<Predicate, Iterator>& rhs)
+    {
+        return lhs.equal(rhs);
+    }
+
+    template <class Predicate, class Iterator>
+    enable_random_access_iterator<Iterator, bool>
+    operator<(const filter_iterator<Predicate, Iterator>& lhs,
+              const filter_iterator<Predicate, Iterator>& rhs)
+    {
+        return lhs.less_than(rhs);
+    }
+
+    template <class Predicate, class Iterator>
+    filter_iterator<std::decay_t<Predicate>, Iterator>
+    make_filter_iterator(Predicate&& pred, Iterator iter, Iterator begin, Iterator end)
+    {
+        using return_type = filter_iterator<std::decay_t<Predicate>, Iterator>;
+        return return_type(
+            std::forward<Predicate>(pred),
+            iter,
+            begin,
+            end);
+    }
+
+    template <class Predicate, class C>
+    filter_iterator<std::decay_t<Predicate>, typename C::iterator>
+    filter_begin(Predicate&& pred, C& container)
+    {
+        using return_type = filter_iterator<std::decay_t<Predicate>, typename C::iterator>;
+        return return_type(
+            std::forward<Predicate>(pred),
+            container.begin(),
+            container.begin(),
+            container.end()
+        );
+    }
+
+    template <class Predicate, class C>
+    filter_iterator<std::decay_t<Predicate>, typename C::iterator>
+    filter_end(Predicate&& pred, C& container)
```

Comment:
I'm not totally sure what's the need for `filter_begin` and `filter_end`. 
I would just write a class for example `filtered<C>` where C is a range. Then the constructor would take a range and a predicate and expose a range interface. That way users can easilly obtain the begin and end filtered, but lazilly (dont go through the range if you dont need the end immediately).
That way (and benefiting from CTAD) users can just use `filtered{ mycontainer, some_predicate}` and use that as a range too.

Reply:
The idea behind `filter_begin` and `filter_end` is to be able to write one liner like `std::copy(filter_begin(pred, container), filter_end(pred, conainer), dst.begin())`.

I agree the repetition is not ideal and can be avoided with the solution you propose, but this requires to instantiate the filter before using it.